CBPs/cbp-0002.md

```diff
@@ -0,0 +1,632 @@
+
+3. ERC777
+    - use send function to repalce transfer function
+    - use the hook function to controls whether to accept transfer
+    - use the operator authorization function to replace the approve function as a proxy transfer solution
+
+#### Improved standards
+
+```
+service: {
+  name: () -> (text) query;
+  symbol: () -> (text) query;
+  decimals: () -> (nat8) query;
+  totalSupply: () -> (nat) query;
+
+  balanceOf: (owner: principal) -> (nat) query;
+  allowance: (owner: principal, spender: principal) -> (nat) query;
+  approve: (spender: principal, value: nat) -> (bool);
+  transferFrom: (sender: principal, receiver: principal, value: nat) -> (bool);
+  transfer: (receiver: principal, value: nat, args:opt vec nat8) -> (bool);
+}
+```
+
+## Suitable for Dfinity
+
+### Problems to be solved
+
+The design of Token Standard will consider the difference between Dfinity and Ethereum fully, and clarify the problems to be solved:
+
+1. No atomicity similar to EVM cross-contract calls
+
+- Solution: [sagas](https://www.cs.cornell.edu/andru/cs711/2002fa/reading/sagas.pdf) is a good solution
+
+2. No built-in event-emit support similar to EVM
+
+- Question: 
+  - How do others know that the transaction happened？
+  - How to check transaction history？
+
+- Consideration: There are two ideas (Pubsub/Notify) for question 1 on the forum
+
+  - When the Token is transferred, Notify informs the recipient, which can fill the missing EVENT. But When the Token recipient not a canister, which means can not notify, it is necessary to support query transaction history.
+
+  - Token does not have sufficient reason to implement Pub/Sub to release transaction information to a third party that is not the recipient
+
+- Solution: 
+  - Notify is a better way; 
+  - It's necessary to support query transaction history;
+
+3. Built-in storage support
+
+- Question: The current storage increase to 8G, but the memory storage is still 4G
+
+- Consideration:
+
+  - Tx history should support query
+
+  - Built-in storage can support Token to store more self-describing information
+
+- Solution:
+
+  - Auto-scaling storage solution
+  - Token implements self-description
+
+4. Reverse gas model
+ - Question: How to prevent ddos attacks
+ - Consideration: The call of the canister does not require the caller to pay gas fees(canister pay the gas fees)
+
+- Solution: Any update call to canister should be charged to the caller
+
+5. There are two different identities in Dfinity, Internet Identity (II for short) and Principal ID
+
+- Question: Which identity to use as a token standard user？
+
+- Consideration: Dfinity's II is an implementation of DID, and II is based on Principal ID
+
+- Solution: It is necessary for the Token standard to be compatible with both identity, in order to meet the needs of different identity scenarios
+
+6. No black hole address
+
+- Question: If there is a need to destroy Token, how to deal with it?
```

Comment:
Something I liked about the ICP design was there was a constant finite number of tokens.
This meant we could use fixed length numbers since overflow was impossible and we could test this invariant.
Having a mint/black hole address solves this.

Reply:
Supporting mint/burn is clear and straightforward for meeting such a need, but not every token needs an interface like mint/burn, so mint/burn should not be in the standard. Tokens that need them implement them themselves

Reply:
ICP is a specific token and therefore can be defined in a targeted manner. However, as a token standard, different token issuers need to be considered, with different requirements for the numbers
> Something I liked about the ICP design was there was a constant finite number of tokens.

Reply:
Fair point